src/TTextEdit.cpp

```diff
@@ -460,43 +466,69 @@ void TTextEdit::drawLine(QPainter& painter, int lineNumber, int lineOfScreen, in
+    QVector<QColor> fgColors;
+    QVector<QRect> textRects;
+    QVector<int> charWidths;
+    QVector<QString> graphemes;
     for (int indexOfChar = 0, total = lineText.size(); indexOfChar < total;) {
         int nextBoundary = boundaryFinder.toNextBoundary();
 
         TChar& charStyle = mpBuffer->buffer.at(lineNumber).at(indexOfChar);
-        int graphemeWidth = drawGrapheme(painter, cursor, lineText.mid(indexOfChar, nextBoundary - indexOfChar), columnWithOutTimestamp, charStyle);
+        int graphemeWidth = drawGraphemeBackground(painter, fgColors, textRects, graphemes, charWidths, cursor, lineText.mid(indexOfChar, nextBoundary - indexOfChar), columnWithOutTimestamp, charStyle);
         cursor.setX(cursor.x() + graphemeWidth);
         indexOfChar = nextBoundary;
         columnWithOutTimestamp += graphemeWidth;
     }
+    boundaryFinder.toStart();
+    int index = -1;
+    for (int indexOfChar = 0, total = lineText.size(); indexOfChar < total;) {
+        int nextBoundary = boundaryFinder.toNextBoundary();
+
+        TChar& charStyle = mpBuffer->buffer.at(lineNumber).at(indexOfChar);
+        ++index;
+        drawGraphemeForeground(painter, fgColors.at(index), textRects.at(index), graphemes.at(index), charStyle);
+        indexOfChar = nextBoundary;
+    }
 }
 
-/**
- * @brief TTextEdit::drawGrapheme
- * @param painter
- * @param cursor
- * @param grapheme
- * @param column Used to calculate the width of Tab
- * @param charStyle
- * @return Return the display width of the grapheme
- */
-int TTextEdit::drawGrapheme(QPainter& painter, const QPoint& cursor, const QString& grapheme, int column, TChar& charStyle) const
+int TTextEdit::drawGraphemeBackground(QPainter& painter, QVector<QColor>& fgColors, QVector<QRect>& textRects, QVector<QString>& graphemes, QVector<int>& charWidths, QPoint& cursor, const QString& grapheme, const int column, TChar& charStyle) const
 {
     static const QString replacementCharacter{QChar::ReplacementCharacter};
     uint unicode = getGraphemeBaseCharacter(grapheme);
     int charWidth;
     bool useReplacementCharacter = false;
     if (unicode == '\t') {
         charWidth = mTabStopwidth - (column % mTabStopwidth);
+        graphemes.append(QString(QChar::Tabulation));
     } else {
         charWidth = getGraphemeWidth(unicode);
         if (!charWidth) {
             // Print the grapheme replacement character instead - which seems to
             // be 1 wide
             useReplacementCharacter = true;
+            charWidth = 1;
         }
+        graphemes.append(useReplacementCharacter ? replacementCharacter : grapheme);
     }
+    charWidths.append(charWidth);
 
+    TChar::AttributeFlags attributes = charStyle.allDisplayAttributes();
+    auto textRect = QRect(mFontWidth * cursor.x(), mFontHeight * cursor.y(), mFontWidth * charWidth, mFontHeight);
+    textRects.append(textRect);
+    QColor bgColor;
+    if (static_cast<bool>(attributes & TChar::Reverse) != charStyle.isSelected()) {
```

Comment:
Missed in current code, but:

```suggestion
    if (Q_UNLIKELY(static_cast<bool>(attributes & TChar::Reverse) != charStyle.isSelected())) {
```

Reply:
How unlikely *is* `Q_UNLIKELY` - what proportion of the time is the **reverse video mode** *flag* ***not*** going to be in the same state as the `(bool) isSelected`?

As it is the two branches of the `if` are of equivalent complexity so it isn't that obvious to me that either needs to be optimised more than the other...

Reply:
I think this is some confusing code that could use clarification then! Are you saying there's an equal chance of someone using the quite exotic reverse ANSI option as someone not using it?

Reply:
Well more that it being used OR text being *selected* but NOT BOTH at the same time... Ah! Yeah, perhaps you are right after all. 😮